workers/unity/Packages/io.improbable.gdk.tools/.CodeGenTemplate/CodeGen/EntryPoint.cs

```diff
@@ -43,27 +49,88 @@ public CodeGenerator(CodeGeneratorOptions options, IFileSystem fileSystem)
         {
             this.options = options;
             this.fileSystem = fileSystem;
+
+            ConfigureLogger();
+            logger = LogManager.GetCurrentClassLogger();
+        }
+
+        private static void ConfigureLogger()
+        {
+            var codeGeneratorOptions =  CodeGeneratorOptions.Instance;
+
+            var config = new NLog.Config.LoggingConfiguration();
+
+            var jsonLayout = new JsonLayout
+            {
+                Attributes =
+                {
+                    new JsonAttribute("time", "${longdate}"),
+                    new JsonAttribute("level", "${level:uppercase=true}"),
+                    new JsonAttribute("logger", "${logger}"),
+                    new JsonAttribute("message", "${message}"),
+                    new JsonAttribute("exception", "${exception:format=ToString}")
+                }
+            };
+
+            var minimumLogLevel = codeGeneratorOptions.DisableVerboseLogging ? LogLevel.Info : LogLevel.Trace;
+
+            if (codeGeneratorOptions.EnableLoggingToConsole)
+            {
+                var consoleTarget = new ConsoleTarget("consoleTarget")
+                {
+                    Layout = jsonLayout
+                };
+                config.AddTarget(consoleTarget);
+                config.AddRule(minimumLogLevel, LogLevel.Fatal, consoleTarget);
+            }
+
+            var fileTarget = new FileTarget("fileTarget")
+            {
+                FileName = codeGeneratorOptions.AbsoluteLogPath,
+                Layout = jsonLayout,
+                DeleteOldFileOnStartup = true
+            };
+            config.AddTarget(fileTarget);
+            config.AddRule(minimumLogLevel, LogLevel.Fatal, fileTarget);
+
+            LogManager.Configuration = config;
         }
 
         public int Run()
         {
+            logger.Info("Starting code generation");
             if (options.ShouldShowHelp)
             {
                 ShowHelpMessage();
                 return 0;
             }
 
-            if (!ValidateOptions())
+            logger.Info("Validating options");
+            var optionErrors = options.GetValidationErrors().ToList();
+            foreach (var optionError in optionErrors)
+            {
+                Console.WriteLine(optionError);
+            }
+
+            if (optionErrors.Any())
             {
                 ShowHelpMessage();
                 return 1;
             }
```

Comment:
I don't think we should log anything until we've done this validation. If my CLI command is incorrect, I don't want to see JSON messages before the actually useful bits